src/NumericalAlgorithms/FiniteDifference/Reconstruct.tpp

```diff
@@ -0,0 +1,267 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "NumericalAlgorithms/FiniteDifference/Reconstruct.hpp"
+
+#include <cstddef>
+
+#include "DataStructures/Index.hpp"
+#include "DataStructures/StripeIterator.hpp"
+#include "DataStructures/Transpose.hpp"
+#include "Domain/Structure/Direction.hpp"
+#include "Domain/Structure/DirectionMap.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace fd::reconstruction::detail {
+template <typename Reconstructor, typename... ArgsForReconstructor, size_t Dim>
+void reconstruct_impl(
+    const gsl::not_null<gsl::span<double>*> recons_upper,
+    const gsl::not_null<gsl::span<double>*> recons_lower,
+    const gsl::span<const double>& volume_vars,
+    const gsl::span<const double>& lower_ghost_data,
+    const gsl::span<const double>& upper_ghost_data,
+    const Index<Dim>& volume_extents, const size_t number_of_variables,
+    const ArgsForReconstructor&... args_for_reconstructor) noexcept {
+  constexpr size_t stencil_width = Reconstructor::stencil_width();
+  ASSERT(stencil_width % 2 == 1, "The stencil with should be odd but got "
+                                     << stencil_width
+                                     << " for the reconstructor.");
+  const size_t ghost_zone_for_stencil = (stencil_width - 1) / 2;
+  // Assume we send one extra ghost cell so we can reconstruct our neighbor's
+  // external data.
```

Comment:
Okay, so here's the situation in art:
```
 Self            Neighbor
x x x x x x | o o
```
if we want to reconstruct the right side of the interface (denoted `|`) then we need to have enough ghost data to be able to actually reconstruct the neighbor (left-most `o`). For minmod/MC this means we need to send 2 ghost cells rather than just 1. For AOWENO53 we need to send 3 instead of 2. We need to do this reconstruction to avoid a second communication. While reconstruction is expensive, I expect it to be cheaper than sending data a second time, but we might want to test that theory at some point in the future :)

Reply:
Got it - thanks!